{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Annotate single cells by reference transcriptomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os,sys\n",
    "import datetime"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import scanpy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scanpy==1.3.4 anndata==0.6.13 numpy==1.15.4 scipy==1.1.0 pandas==0.23.4 scikit-learn==0.20.1 statsmodels==0.9.0 python-igraph==0.7.1 louvain==0.6.1 \n",
      "Memory usage: current 0.17 GB, difference +0.17 GB\n"
     ]
    }
   ],
   "source": [
    "import scanpy.api as sc\n",
    "sc.logging.print_versions()\n",
    "sc.logging.print_memory_usage()\n",
    "sc.settings.verbosity = 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import my utility functions and import statements from github"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'# This cell is run once to download my custom functions and import statements from github\\n\\n!git clone --depth=1 https://github.com/rapolaszilionis/utility_functions\\n    \\n# github doesn\\'t seem to have an option to download a specific version of the repo from the history.\\n# So I download my utility functions and save the download time by appending it to the directory name.\\n# These utility functions to be shared together with the notebook.\\n\\ntoappend = datetime.datetime.now().strftime(\\'%y%m%d_%Hh%M\\')\\nnewname = \"utility_functions_%s\"%toappend\\nprint(newname)\\n\\n\\n# rename the py file with utility functions\\nos.rename(\"utility_functions\",newname)'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"# This cell is run once to download my custom functions and import statements from github\n",
    "\n",
    "!git clone --depth=1 https://github.com/rapolaszilionis/utility_functions\n",
    "    \n",
    "# github doesn't seem to have an option to download a specific version of the repo from the history.\n",
    "# So I download my utility functions and save the download time by appending it to the directory name.\n",
    "# These utility functions to be shared together with the notebook.\n",
    "\n",
    "toappend = datetime.datetime.now().strftime('%y%m%d_%Hh%M')\n",
    "newname = \"utility_functions_%s\"%toappend\n",
    "print(newname)\n",
    "\n",
    "\n",
    "# rename the py file with utility functions\n",
    "os.rename(\"utility_functions\",newname)\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "python version: 3.6.7\n"
     ]
    }
   ],
   "source": [
    "# add the utility function folder to PATH\n",
    "sys.path.append(os.path.abspath(\"utility_functions_190425_11h32\"))\n",
    "\n",
    "from rz_import_statements import *\n",
    "import rz_functions as rz\n",
    "import rz_fig_params # this adjust mpl.rcParams, almost nothing to import\n",
    "import rz_utility_spring as srz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load data and place into an annData object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cell_info_8594x25.tsv         injury_barcodes.tsv\r\n",
      "control_barcodes.tsv          injury_genes.tsv\r\n",
      "control_genes.tsv             injury_loom.loom\r\n",
      "control_loom.loom             injury_matrix.mtx\r\n",
      "control_matrix.mtx            raw_counts_8594x27998.mtx\r\n",
      "gene_names_alphabetically.txt\r\n"
     ]
    }
   ],
   "source": [
    "!ls data_from_geo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "obspath = 'data_from_geo/cell_info_8594x25.tsv'\n",
    "countpath = 'data_from_geo/raw_counts_8594x27998.mtx'\n",
    "genepath = 'data_from_geo/gene_names_alphabetically.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(8594, 25)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>condition</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>pass_quality_filters</th>\n",
       "      <th>inj_epithelial</th>\n",
       "      <th>ctr_epithelial</th>\n",
       "      <th>excluded_as_immune_or_mesench</th>\n",
       "      <th>class</th>\n",
       "      <th>population</th>\n",
       "      <th>phase</th>\n",
       "      <th>...</th>\n",
       "      <th>x_control</th>\n",
       "      <th>y_control</th>\n",
       "      <th>x_class3_exploded</th>\n",
       "      <th>y_class3_exploded</th>\n",
       "      <th>x_control_injured</th>\n",
       "      <th>y_control_injured</th>\n",
       "      <th>x_class1_CTR_cell_cyc_removed</th>\n",
       "      <th>y_class1_CTR_cell_cyc_removed</th>\n",
       "      <th>x_class1_INJ_cell_cyc_removed</th>\n",
       "      <th>y_class1_INJ_cell_cyc_removed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCTGAGTGCTGCC-1</td>\n",
       "      <td>control</td>\n",
       "      <td>2787</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>class3</td>\n",
       "      <td>ctr_DEEx</td>\n",
       "      <td>G1</td>\n",
       "      <td>...</td>\n",
       "      <td>783.705500</td>\n",
       "      <td>-377.074709</td>\n",
       "      <td>1082.172898</td>\n",
       "      <td>-266.303304</td>\n",
       "      <td>709.78614</td>\n",
       "      <td>-325.505019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCTGAGTGGGTTG-1</td>\n",
       "      <td>control</td>\n",
       "      <td>3325</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>class3</td>\n",
       "      <td>ctr_upper_IEE</td>\n",
       "      <td>G1</td>\n",
       "      <td>...</td>\n",
       "      <td>709.612642</td>\n",
       "      <td>-429.409941</td>\n",
       "      <td>854.230866</td>\n",
       "      <td>-454.025809</td>\n",
       "      <td>529.15516</td>\n",
       "      <td>-471.961493</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCTGCAAGTCTAC-1</td>\n",
       "      <td>control</td>\n",
       "      <td>1781</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCTGCAATCTGCA-1</td>\n",
       "      <td>control</td>\n",
       "      <td>3468</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCTGCACGGTGTC-1</td>\n",
       "      <td>control</td>\n",
       "      <td>1745</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              barcode condition  total_counts  pass_quality_filters  \\\n",
       "0  AAACCTGAGTGCTGCC-1   control          2787                  True   \n",
       "1  AAACCTGAGTGGGTTG-1   control          3325                  True   \n",
       "2  AAACCTGCAAGTCTAC-1   control          1781                 False   \n",
       "3  AAACCTGCAATCTGCA-1   control          3468                  True   \n",
       "4  AAACCTGCACGGTGTC-1   control          1745                 False   \n",
       "\n",
       "   inj_epithelial  ctr_epithelial  excluded_as_immune_or_mesench   class  \\\n",
       "0           False            True                          False  class3   \n",
       "1           False            True                          False  class3   \n",
       "2           False           False                          False     NaN   \n",
       "3           False           False                           True     NaN   \n",
       "4           False           False                          False     NaN   \n",
       "\n",
       "      population phase              ...                 x_control   y_control  \\\n",
       "0       ctr_DEEx    G1              ...                783.705500 -377.074709   \n",
       "1  ctr_upper_IEE    G1              ...                709.612642 -429.409941   \n",
       "2            NaN   NaN              ...                       NaN         NaN   \n",
       "3            NaN   NaN              ...                       NaN         NaN   \n",
       "4            NaN   NaN              ...                       NaN         NaN   \n",
       "\n",
       "  x_class3_exploded  y_class3_exploded x_control_injured  y_control_injured  \\\n",
       "0       1082.172898        -266.303304         709.78614        -325.505019   \n",
       "1        854.230866        -454.025809         529.15516        -471.961493   \n",
       "2               NaN                NaN               NaN                NaN   \n",
       "3               NaN                NaN               NaN                NaN   \n",
       "4               NaN                NaN               NaN                NaN   \n",
       "\n",
       "   x_class1_CTR_cell_cyc_removed  y_class1_CTR_cell_cyc_removed  \\\n",
       "0                            NaN                            NaN   \n",
       "1                            NaN                            NaN   \n",
       "2                            NaN                            NaN   \n",
       "3                            NaN                            NaN   \n",
       "4                            NaN                            NaN   \n",
       "\n",
       "   x_class1_INJ_cell_cyc_removed  y_class1_INJ_cell_cyc_removed  \n",
       "0                            NaN                            NaN  \n",
       "1                            NaN                            NaN  \n",
       "2                            NaN                            NaN  \n",
       "3                            NaN                            NaN  \n",
       "4                            NaN                            NaN  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load cell info\n",
    "obs = pd.read_csv(obspath,sep='\\t',index_col=0)\n",
    "print(obs.shape)\n",
    "obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(8594, 27998)\n"
     ]
    }
   ],
   "source": [
    "# load counts\n",
    "adata = sc.read_mtx(countpath)\n",
    "print(adata.shape)\n",
    "\n",
    "# add genes (annotation of variables)\n",
    "adata.var['genes'] = np.loadtxt(genepath,dtype=str)\n",
    "\n",
    "# make sure var names are genes\n",
    "adata.var_names = adata.var['genes'].values\n",
    "\n",
    "# add obs (annotation of observations)\n",
    "adata.obs = obs\n",
    "\n",
    "# make sure index is unique AND a string\n",
    "adata.obs_names_make_unique()\n",
    "adata.obs_names = adata.obs_names.astype(str)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Scale (normalize) data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[2787.]\n",
      " [3325.]\n",
      " [1781.]\n",
      " [3468.]\n",
      " [1745.]]\n",
      "\n",
      "[[10000.   ]\n",
      " [10000.   ]\n",
      " [10000.001]\n",
      " [10000.   ]\n",
      " [ 9999.999]]\n"
     ]
    }
   ],
   "source": [
    "# turn into counts per 10k\n",
    "print(adata.X[:5,:].sum(axis=1))\n",
    "print()\n",
    "sc.pp.normalize_per_cell(adata, counts_per_cell_after=1e4)\n",
    "print(adata.X[:5,:].sum(axis=1))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get centroids (reference transcriptomes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function centroids in module rz_functions:\n",
      "\n",
      "centroids(label, adata, E=None, gene_list=None)\n",
      "    Calculate average gene expression level per cell label (e.g. cluster).\n",
      "    input:\n",
      "        - label: name of column that stores the label of interest in adata.obs\n",
      "        - adata: AnnData object OR a cell x feature pandas dataframe with label as one of the columns\n",
      "        - E and gene_list: optional and only used when adata is not an AnnData object. In that case\n",
      "        the cells x genes sparse expression matrix E and the gene_list must be specified\n",
      "        \n",
      "    returns:\n",
      "        pandas dataframe, centroids x genes\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(rz.centroids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "label = 'population'\n",
    "cell_mask = adata.obs['ctr_epithelial'].astype(bool).values\n",
    "centroids = rz.centroids(label,adata[cell_mask])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0610007P14Rik</th>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <th>0610009L18Rik</th>\n",
       "      <th>0610009O20Rik</th>\n",
       "      <th>0610010F05Rik</th>\n",
       "      <th>0610010K14Rik</th>\n",
       "      <th>0610011F06Rik</th>\n",
       "      <th>0610012D04Rik</th>\n",
       "      <th>0610012G03Rik</th>\n",
       "      <th>0610025J13Rik</th>\n",
       "      <th>...</th>\n",
       "      <th>mt-Co2</th>\n",
       "      <th>mt-Co3</th>\n",
       "      <th>mt-Cytb</th>\n",
       "      <th>mt-Nd1</th>\n",
       "      <th>mt-Nd2</th>\n",
       "      <th>mt-Nd3</th>\n",
       "      <th>mt-Nd4</th>\n",
       "      <th>mt-Nd4l</th>\n",
       "      <th>mt-Nd5</th>\n",
       "      <th>mt-Nd6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ctr_DEEx</th>\n",
       "      <td>1.391852</td>\n",
       "      <td>0.879937</td>\n",
       "      <td>0.324726</td>\n",
       "      <td>0.209586</td>\n",
       "      <td>0.086968</td>\n",
       "      <td>0.329960</td>\n",
       "      <td>0.954454</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.200782</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>14.079573</td>\n",
       "      <td>83.904465</td>\n",
       "      <td>56.074482</td>\n",
       "      <td>23.852005</td>\n",
       "      <td>3.198850</td>\n",
       "      <td>1.030915</td>\n",
       "      <td>18.962263</td>\n",
       "      <td>0.987729</td>\n",
       "      <td>1.261443</td>\n",
       "      <td>0.275973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctr_upper_IEE</th>\n",
       "      <td>1.300044</td>\n",
       "      <td>0.695099</td>\n",
       "      <td>0.314438</td>\n",
       "      <td>0.287892</td>\n",
       "      <td>0.184926</td>\n",
       "      <td>0.397997</td>\n",
       "      <td>0.773451</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.998683</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>15.316485</td>\n",
       "      <td>84.531807</td>\n",
       "      <td>54.308296</td>\n",
       "      <td>24.424025</td>\n",
       "      <td>3.601944</td>\n",
       "      <td>1.055761</td>\n",
       "      <td>18.239050</td>\n",
       "      <td>0.895775</td>\n",
       "      <td>0.950987</td>\n",
       "      <td>0.232046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctr_M_G1</th>\n",
       "      <td>1.303535</td>\n",
       "      <td>0.552314</td>\n",
       "      <td>0.065731</td>\n",
       "      <td>0.330279</td>\n",
       "      <td>0.126810</td>\n",
       "      <td>0.566542</td>\n",
       "      <td>0.606099</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.455639</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>12.348806</td>\n",
       "      <td>70.209267</td>\n",
       "      <td>47.538788</td>\n",
       "      <td>23.879992</td>\n",
       "      <td>2.751078</td>\n",
       "      <td>0.998057</td>\n",
       "      <td>15.154018</td>\n",
       "      <td>0.907756</td>\n",
       "      <td>0.761977</td>\n",
       "      <td>0.202841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctr_OEE_IEE</th>\n",
       "      <td>1.737094</td>\n",
       "      <td>0.544913</td>\n",
       "      <td>0.267885</td>\n",
       "      <td>0.461709</td>\n",
       "      <td>0.105900</td>\n",
       "      <td>0.447238</td>\n",
       "      <td>0.927448</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.360467</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>13.681261</td>\n",
       "      <td>85.969254</td>\n",
       "      <td>55.733368</td>\n",
       "      <td>22.308928</td>\n",
       "      <td>4.568748</td>\n",
       "      <td>0.957020</td>\n",
       "      <td>19.031935</td>\n",
       "      <td>0.749691</td>\n",
       "      <td>1.071926</td>\n",
       "      <td>0.076110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctr_G2_M</th>\n",
       "      <td>2.209283</td>\n",
       "      <td>0.774067</td>\n",
       "      <td>0.063674</td>\n",
       "      <td>0.473809</td>\n",
       "      <td>0.182266</td>\n",
       "      <td>0.590541</td>\n",
       "      <td>0.573025</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.316742</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>12.441298</td>\n",
       "      <td>70.319290</td>\n",
       "      <td>45.029736</td>\n",
       "      <td>23.863428</td>\n",
       "      <td>3.214058</td>\n",
       "      <td>0.953022</td>\n",
       "      <td>15.301864</td>\n",
       "      <td>0.760218</td>\n",
       "      <td>1.103230</td>\n",
       "      <td>0.170797</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 27998 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               0610007P14Rik  0610009B22Rik  0610009L18Rik  0610009O20Rik  \\\n",
       "ctr_DEEx            1.391852       0.879937       0.324726       0.209586   \n",
       "ctr_upper_IEE       1.300044       0.695099       0.314438       0.287892   \n",
       "ctr_M_G1            1.303535       0.552314       0.065731       0.330279   \n",
       "ctr_OEE_IEE         1.737094       0.544913       0.267885       0.461709   \n",
       "ctr_G2_M            2.209283       0.774067       0.063674       0.473809   \n",
       "\n",
       "               0610010F05Rik  0610010K14Rik  0610011F06Rik  0610012D04Rik  \\\n",
       "ctr_DEEx            0.086968       0.329960       0.954454            0.0   \n",
       "ctr_upper_IEE       0.184926       0.397997       0.773451            0.0   \n",
       "ctr_M_G1            0.126810       0.566542       0.606099            0.0   \n",
       "ctr_OEE_IEE         0.105900       0.447238       0.927448            0.0   \n",
       "ctr_G2_M            0.182266       0.590541       0.573025            0.0   \n",
       "\n",
       "               0610012G03Rik  0610025J13Rik    ...        mt-Co2     mt-Co3  \\\n",
       "ctr_DEEx            1.200782            0.0    ...     14.079573  83.904465   \n",
       "ctr_upper_IEE       0.998683            0.0    ...     15.316485  84.531807   \n",
       "ctr_M_G1            1.455639            0.0    ...     12.348806  70.209267   \n",
       "ctr_OEE_IEE         1.360467            0.0    ...     13.681261  85.969254   \n",
       "ctr_G2_M            1.316742            0.0    ...     12.441298  70.319290   \n",
       "\n",
       "                 mt-Cytb     mt-Nd1    mt-Nd2    mt-Nd3     mt-Nd4   mt-Nd4l  \\\n",
       "ctr_DEEx       56.074482  23.852005  3.198850  1.030915  18.962263  0.987729   \n",
       "ctr_upper_IEE  54.308296  24.424025  3.601944  1.055761  18.239050  0.895775   \n",
       "ctr_M_G1       47.538788  23.879992  2.751078  0.998057  15.154018  0.907756   \n",
       "ctr_OEE_IEE    55.733368  22.308928  4.568748  0.957020  19.031935  0.749691   \n",
       "ctr_G2_M       45.029736  23.863428  3.214058  0.953022  15.301864  0.760218   \n",
       "\n",
       "                 mt-Nd5    mt-Nd6  \n",
       "ctr_DEEx       1.261443  0.275973  \n",
       "ctr_upper_IEE  0.950987  0.232046  \n",
       "ctr_M_G1       0.761977  0.202841  \n",
       "ctr_OEE_IEE    1.071926  0.076110  \n",
       "ctr_G2_M       1.103230  0.170797  \n",
       "\n",
       "[5 rows x 27998 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "centroids.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run Bayesian classifier using all genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# select cells to classify\n",
    "toclassify = adata.obs['inj_epithelial'].astype(bool).values\n",
    "E = adata[toclassify].X\n",
    "gene_list = adata.var_names\n",
    "\n",
    "# reference transcriptomes to classify by\n",
    "categories = centroids.T\n",
    "\n",
    "# Select genes. They have to be present in both datasets.\n",
    "genes_to_use = gene_list #in this case simply all genes\n",
    "gene_mask = np.in1d(gene_list,genes_to_use)\n",
    "\n",
    "pseudo = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.02 min.\n",
      "cells from 0 to 1000 done\n",
      "0.04 min.\n",
      "cells from 1000 to 2000 done\n",
      "0.06 min.\n",
      "cells from 2000 to 3000 done\n",
      "0.08 min.\n",
      "cells from 3000 to 4000 done\n",
      "0.09 min.\n",
      "cells from 4000 to 4344 done\n"
     ]
    }
   ],
   "source": [
    "start = time.time()\n",
    "bays = []\n",
    "i = 0\n",
    "interval = 1000\n",
    "for j in range(interval,E.shape[0]+interval,interval):\n",
    "    j = min(j,E.shape[0])\n",
    "    stepsize = j-i\n",
    "    tmp_dense = pd.DataFrame(E.T[gene_mask][:,i:j].todense())\n",
    "    tmp_dense.index = np.array(gene_list)[gene_mask]\n",
    "    bay = rz.bayesian_classifier(tmp_dense,categories.loc[tmp_dense.index]+pseudo)\n",
    "    bays.append(bay)\n",
    "    i = j\n",
    "    print('%.2f min.'%((time.time()-start)/60.))\n",
    "    print('cells from %d to %d done'%(i-stepsize,min(i,E.shape[0])))\n",
    "    \n",
    "bay = pd.concat(bays,axis=1)\n",
    "bay.columns = np.arange(bay.shape[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>4334</th>\n",
       "      <th>4335</th>\n",
       "      <th>4336</th>\n",
       "      <th>4337</th>\n",
       "      <th>4338</th>\n",
       "      <th>4339</th>\n",
       "      <th>4340</th>\n",
       "      <th>4341</th>\n",
       "      <th>4342</th>\n",
       "      <th>4343</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ctr_DEEx</th>\n",
       "      <td>-37827.351562</td>\n",
       "      <td>-37193.226562</td>\n",
       "      <td>-37724.312500</td>\n",
       "      <td>-38271.964844</td>\n",
       "      <td>-38053.949219</td>\n",
       "      <td>-38834.886719</td>\n",
       "      <td>-36969.839844</td>\n",
       "      <td>-38269.835938</td>\n",
       "      <td>-38657.800781</td>\n",
       "      <td>-37963.593750</td>\n",
       "      <td>...</td>\n",
       "      <td>-37493.234375</td>\n",
       "      <td>-37796.992188</td>\n",
       "      <td>-38086.953125</td>\n",
       "      <td>-38499.703125</td>\n",
       "      <td>-37928.050781</td>\n",
       "      <td>-37438.015625</td>\n",
       "      <td>-37693.460938</td>\n",
       "      <td>-38000.359375</td>\n",
       "      <td>-37698.652344</td>\n",
       "      <td>-37769.933594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctr_upper_IEE</th>\n",
       "      <td>-37779.187500</td>\n",
       "      <td>-37301.761719</td>\n",
       "      <td>-37749.914062</td>\n",
       "      <td>-38292.160156</td>\n",
       "      <td>-37981.261719</td>\n",
       "      <td>-38974.894531</td>\n",
       "      <td>-37144.335938</td>\n",
       "      <td>-38224.960938</td>\n",
       "      <td>-38696.679688</td>\n",
       "      <td>-38067.812500</td>\n",
       "      <td>...</td>\n",
       "      <td>-37432.890625</td>\n",
       "      <td>-37804.593750</td>\n",
       "      <td>-38370.593750</td>\n",
       "      <td>-38586.027344</td>\n",
       "      <td>-38070.097656</td>\n",
       "      <td>-37576.816406</td>\n",
       "      <td>-37768.582031</td>\n",
       "      <td>-38050.085938</td>\n",
       "      <td>-37972.800781</td>\n",
       "      <td>-37887.433594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctr_M_G1</th>\n",
       "      <td>-37633.441406</td>\n",
       "      <td>-37372.425781</td>\n",
       "      <td>-37515.351562</td>\n",
       "      <td>-38080.683594</td>\n",
       "      <td>-37812.203125</td>\n",
       "      <td>-38638.175781</td>\n",
       "      <td>-37195.031250</td>\n",
       "      <td>-38118.628906</td>\n",
       "      <td>-38419.722656</td>\n",
       "      <td>-37928.214844</td>\n",
       "      <td>...</td>\n",
       "      <td>-37408.769531</td>\n",
       "      <td>-37672.359375</td>\n",
       "      <td>-38292.304688</td>\n",
       "      <td>-38289.351562</td>\n",
       "      <td>-38093.101562</td>\n",
       "      <td>-37668.308594</td>\n",
       "      <td>-37772.933594</td>\n",
       "      <td>-37999.277344</td>\n",
       "      <td>-37982.156250</td>\n",
       "      <td>-37893.882812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctr_OEE_IEE</th>\n",
       "      <td>-38003.457031</td>\n",
       "      <td>-37286.925781</td>\n",
       "      <td>-37948.328125</td>\n",
       "      <td>-38363.226562</td>\n",
       "      <td>-38153.125000</td>\n",
       "      <td>-38986.945312</td>\n",
       "      <td>-37022.015625</td>\n",
       "      <td>-38382.597656</td>\n",
       "      <td>-38831.328125</td>\n",
       "      <td>-38218.882812</td>\n",
       "      <td>...</td>\n",
       "      <td>-37659.273438</td>\n",
       "      <td>-37983.074219</td>\n",
       "      <td>-38369.308594</td>\n",
       "      <td>-38710.765625</td>\n",
       "      <td>-37998.824219</td>\n",
       "      <td>-37353.781250</td>\n",
       "      <td>-37843.414062</td>\n",
       "      <td>-38128.695312</td>\n",
       "      <td>-37844.609375</td>\n",
       "      <td>-37867.457031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctr_G2_M</th>\n",
       "      <td>-37814.449219</td>\n",
       "      <td>-37498.437500</td>\n",
       "      <td>-37720.675781</td>\n",
       "      <td>-37941.269531</td>\n",
       "      <td>-37914.281250</td>\n",
       "      <td>-38463.660156</td>\n",
       "      <td>-37277.367188</td>\n",
       "      <td>-38132.328125</td>\n",
       "      <td>-38529.675781</td>\n",
       "      <td>-38074.367188</td>\n",
       "      <td>...</td>\n",
       "      <td>-37656.128906</td>\n",
       "      <td>-37866.792969</td>\n",
       "      <td>-38396.371094</td>\n",
       "      <td>-38167.925781</td>\n",
       "      <td>-38156.140625</td>\n",
       "      <td>-37678.015625</td>\n",
       "      <td>-37918.812500</td>\n",
       "      <td>-38147.777344</td>\n",
       "      <td>-38059.562500</td>\n",
       "      <td>-37993.644531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctr_OEE_2</th>\n",
       "      <td>-38277.347656</td>\n",
       "      <td>-37329.871094</td>\n",
       "      <td>-38160.781250</td>\n",
       "      <td>-38607.304688</td>\n",
       "      <td>-38447.250000</td>\n",
       "      <td>-38797.902344</td>\n",
       "      <td>-36679.453125</td>\n",
       "      <td>-38618.253906</td>\n",
       "      <td>-38712.855469</td>\n",
       "      <td>-38347.917969</td>\n",
       "      <td>...</td>\n",
       "      <td>-37911.523438</td>\n",
       "      <td>-38158.085938</td>\n",
       "      <td>-38149.726562</td>\n",
       "      <td>-38825.769531</td>\n",
       "      <td>-37675.179688</td>\n",
       "      <td>-37156.734375</td>\n",
       "      <td>-37710.210938</td>\n",
       "      <td>-38273.039062</td>\n",
       "      <td>-37458.312500</td>\n",
       "      <td>-37666.750000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctr_OSR</th>\n",
       "      <td>-38089.160156</td>\n",
       "      <td>-37278.207031</td>\n",
       "      <td>-38011.726562</td>\n",
       "      <td>-38462.550781</td>\n",
       "      <td>-38229.312500</td>\n",
       "      <td>-38879.316406</td>\n",
       "      <td>-36921.203125</td>\n",
       "      <td>-38346.566406</td>\n",
       "      <td>-38695.714844</td>\n",
       "      <td>-38299.664062</td>\n",
       "      <td>...</td>\n",
       "      <td>-37783.640625</td>\n",
       "      <td>-38048.109375</td>\n",
       "      <td>-38352.976562</td>\n",
       "      <td>-38723.402344</td>\n",
       "      <td>-37816.804688</td>\n",
       "      <td>-37350.921875</td>\n",
       "      <td>-37750.824219</td>\n",
       "      <td>-38068.828125</td>\n",
       "      <td>-37815.625000</td>\n",
       "      <td>-37756.257812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctr_SI</th>\n",
       "      <td>-37922.355469</td>\n",
       "      <td>-37182.781250</td>\n",
       "      <td>-37939.300781</td>\n",
       "      <td>-38443.992188</td>\n",
       "      <td>-38170.445312</td>\n",
       "      <td>-38995.699219</td>\n",
       "      <td>-37008.585938</td>\n",
       "      <td>-38277.964844</td>\n",
       "      <td>-38743.406250</td>\n",
       "      <td>-38198.730469</td>\n",
       "      <td>...</td>\n",
       "      <td>-37566.816406</td>\n",
       "      <td>-37883.609375</td>\n",
       "      <td>-38368.410156</td>\n",
       "      <td>-38707.847656</td>\n",
       "      <td>-37883.156250</td>\n",
       "      <td>-37407.363281</td>\n",
       "      <td>-37587.226562</td>\n",
       "      <td>-37981.140625</td>\n",
       "      <td>-37871.667969</td>\n",
       "      <td>-37822.781250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctr_S</th>\n",
       "      <td>-37689.898438</td>\n",
       "      <td>-37388.378906</td>\n",
       "      <td>-37678.914062</td>\n",
       "      <td>-38042.070312</td>\n",
       "      <td>-37892.906250</td>\n",
       "      <td>-38731.605469</td>\n",
       "      <td>-37208.656250</td>\n",
       "      <td>-38030.496094</td>\n",
       "      <td>-38600.886719</td>\n",
       "      <td>-37993.128906</td>\n",
       "      <td>...</td>\n",
       "      <td>-37540.445312</td>\n",
       "      <td>-37748.078125</td>\n",
       "      <td>-38382.832031</td>\n",
       "      <td>-38282.371094</td>\n",
       "      <td>-38123.441406</td>\n",
       "      <td>-37640.625000</td>\n",
       "      <td>-37844.835938</td>\n",
       "      <td>-38028.097656</td>\n",
       "      <td>-38010.605469</td>\n",
       "      <td>-37947.890625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctr_VEE</th>\n",
       "      <td>-38468.382812</td>\n",
       "      <td>-37540.101562</td>\n",
       "      <td>-38340.515625</td>\n",
       "      <td>-38810.101562</td>\n",
       "      <td>-38606.800781</td>\n",
       "      <td>-38949.019531</td>\n",
       "      <td>-36767.312500</td>\n",
       "      <td>-38805.835938</td>\n",
       "      <td>-38693.207031</td>\n",
       "      <td>-38509.378906</td>\n",
       "      <td>...</td>\n",
       "      <td>-38065.285156</td>\n",
       "      <td>-38296.089844</td>\n",
       "      <td>-38450.222656</td>\n",
       "      <td>-38989.757812</td>\n",
       "      <td>-37540.753906</td>\n",
       "      <td>-37356.140625</td>\n",
       "      <td>-37515.898438</td>\n",
       "      <td>-38469.417969</td>\n",
       "      <td>-37611.816406</td>\n",
       "      <td>-37702.785156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctr_pre_AMB</th>\n",
       "      <td>-37741.093750</td>\n",
       "      <td>-37225.371094</td>\n",
       "      <td>-37829.023438</td>\n",
       "      <td>-38415.410156</td>\n",
       "      <td>-38099.085938</td>\n",
       "      <td>-38979.183594</td>\n",
       "      <td>-37066.386719</td>\n",
       "      <td>-38346.269531</td>\n",
       "      <td>-38732.742188</td>\n",
       "      <td>-38157.570312</td>\n",
       "      <td>...</td>\n",
       "      <td>-37423.406250</td>\n",
       "      <td>-37732.039062</td>\n",
       "      <td>-38350.027344</td>\n",
       "      <td>-38639.386719</td>\n",
       "      <td>-37994.761719</td>\n",
       "      <td>-37521.710938</td>\n",
       "      <td>-37740.382812</td>\n",
       "      <td>-38074.945312</td>\n",
       "      <td>-37939.484375</td>\n",
       "      <td>-37864.484375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctr_OEE_1</th>\n",
       "      <td>-38383.949219</td>\n",
       "      <td>-37353.816406</td>\n",
       "      <td>-38284.984375</td>\n",
       "      <td>-38680.019531</td>\n",
       "      <td>-38574.238281</td>\n",
       "      <td>-39073.683594</td>\n",
       "      <td>-36963.960938</td>\n",
       "      <td>-38691.554688</td>\n",
       "      <td>-39050.984375</td>\n",
       "      <td>-38500.695312</td>\n",
       "      <td>...</td>\n",
       "      <td>-38013.066406</td>\n",
       "      <td>-38286.523438</td>\n",
       "      <td>-38257.335938</td>\n",
       "      <td>-38978.710938</td>\n",
       "      <td>-37981.210938</td>\n",
       "      <td>-37268.257812</td>\n",
       "      <td>-38017.257812</td>\n",
       "      <td>-38365.375000</td>\n",
       "      <td>-37710.820312</td>\n",
       "      <td>-37919.898438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctr_AMB_dist</th>\n",
       "      <td>-39180.464844</td>\n",
       "      <td>-38451.742188</td>\n",
       "      <td>-39382.187500</td>\n",
       "      <td>-39852.933594</td>\n",
       "      <td>-39601.050781</td>\n",
       "      <td>-40060.781250</td>\n",
       "      <td>-38137.855469</td>\n",
       "      <td>-39814.390625</td>\n",
       "      <td>-39892.593750</td>\n",
       "      <td>-39647.953125</td>\n",
       "      <td>...</td>\n",
       "      <td>-38947.949219</td>\n",
       "      <td>-39086.531250</td>\n",
       "      <td>-39551.707031</td>\n",
       "      <td>-40016.859375</td>\n",
       "      <td>-38868.500000</td>\n",
       "      <td>-38637.738281</td>\n",
       "      <td>-38798.398438</td>\n",
       "      <td>-39448.601562</td>\n",
       "      <td>-39047.828125</td>\n",
       "      <td>-39004.890625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctr_AMB_prox</th>\n",
       "      <td>-38604.132812</td>\n",
       "      <td>-37959.253906</td>\n",
       "      <td>-38812.621094</td>\n",
       "      <td>-39376.335938</td>\n",
       "      <td>-39074.160156</td>\n",
       "      <td>-39709.210938</td>\n",
       "      <td>-37730.578125</td>\n",
       "      <td>-39306.699219</td>\n",
       "      <td>-39482.058594</td>\n",
       "      <td>-39121.781250</td>\n",
       "      <td>...</td>\n",
       "      <td>-38368.785156</td>\n",
       "      <td>-38536.359375</td>\n",
       "      <td>-39138.968750</td>\n",
       "      <td>-39561.261719</td>\n",
       "      <td>-38505.507812</td>\n",
       "      <td>-38302.078125</td>\n",
       "      <td>-38372.339844</td>\n",
       "      <td>-38936.535156</td>\n",
       "      <td>-38693.031250</td>\n",
       "      <td>-38571.542969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctr_ISR_SI</th>\n",
       "      <td>-38149.710938</td>\n",
       "      <td>-37224.031250</td>\n",
       "      <td>-38083.921875</td>\n",
       "      <td>-38603.601562</td>\n",
       "      <td>-38379.433594</td>\n",
       "      <td>-38899.046875</td>\n",
       "      <td>-36920.953125</td>\n",
       "      <td>-38447.234375</td>\n",
       "      <td>-38797.140625</td>\n",
       "      <td>-38369.179688</td>\n",
       "      <td>...</td>\n",
       "      <td>-37848.394531</td>\n",
       "      <td>-38103.480469</td>\n",
       "      <td>-38193.117188</td>\n",
       "      <td>-38778.691406</td>\n",
       "      <td>-37830.945312</td>\n",
       "      <td>-37364.503906</td>\n",
       "      <td>-37787.656250</td>\n",
       "      <td>-38050.359375</td>\n",
       "      <td>-37776.214844</td>\n",
       "      <td>-37771.726562</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15 rows × 4344 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       0             1             2             3     \\\n",
       "ctr_DEEx      -37827.351562 -37193.226562 -37724.312500 -38271.964844   \n",
       "ctr_upper_IEE -37779.187500 -37301.761719 -37749.914062 -38292.160156   \n",
       "ctr_M_G1      -37633.441406 -37372.425781 -37515.351562 -38080.683594   \n",
       "ctr_OEE_IEE   -38003.457031 -37286.925781 -37948.328125 -38363.226562   \n",
       "ctr_G2_M      -37814.449219 -37498.437500 -37720.675781 -37941.269531   \n",
       "ctr_OEE_2     -38277.347656 -37329.871094 -38160.781250 -38607.304688   \n",
       "ctr_OSR       -38089.160156 -37278.207031 -38011.726562 -38462.550781   \n",
       "ctr_SI        -37922.355469 -37182.781250 -37939.300781 -38443.992188   \n",
       "ctr_S         -37689.898438 -37388.378906 -37678.914062 -38042.070312   \n",
       "ctr_VEE       -38468.382812 -37540.101562 -38340.515625 -38810.101562   \n",
       "ctr_pre_AMB   -37741.093750 -37225.371094 -37829.023438 -38415.410156   \n",
       "ctr_OEE_1     -38383.949219 -37353.816406 -38284.984375 -38680.019531   \n",
       "ctr_AMB_dist  -39180.464844 -38451.742188 -39382.187500 -39852.933594   \n",
       "ctr_AMB_prox  -38604.132812 -37959.253906 -38812.621094 -39376.335938   \n",
       "ctr_ISR_SI    -38149.710938 -37224.031250 -38083.921875 -38603.601562   \n",
       "\n",
       "                       4             5             6             7     \\\n",
       "ctr_DEEx      -38053.949219 -38834.886719 -36969.839844 -38269.835938   \n",
       "ctr_upper_IEE -37981.261719 -38974.894531 -37144.335938 -38224.960938   \n",
       "ctr_M_G1      -37812.203125 -38638.175781 -37195.031250 -38118.628906   \n",
       "ctr_OEE_IEE   -38153.125000 -38986.945312 -37022.015625 -38382.597656   \n",
       "ctr_G2_M      -37914.281250 -38463.660156 -37277.367188 -38132.328125   \n",
       "ctr_OEE_2     -38447.250000 -38797.902344 -36679.453125 -38618.253906   \n",
       "ctr_OSR       -38229.312500 -38879.316406 -36921.203125 -38346.566406   \n",
       "ctr_SI        -38170.445312 -38995.699219 -37008.585938 -38277.964844   \n",
       "ctr_S         -37892.906250 -38731.605469 -37208.656250 -38030.496094   \n",
       "ctr_VEE       -38606.800781 -38949.019531 -36767.312500 -38805.835938   \n",
       "ctr_pre_AMB   -38099.085938 -38979.183594 -37066.386719 -38346.269531   \n",
       "ctr_OEE_1     -38574.238281 -39073.683594 -36963.960938 -38691.554688   \n",
       "ctr_AMB_dist  -39601.050781 -40060.781250 -38137.855469 -39814.390625   \n",
       "ctr_AMB_prox  -39074.160156 -39709.210938 -37730.578125 -39306.699219   \n",
       "ctr_ISR_SI    -38379.433594 -38899.046875 -36920.953125 -38447.234375   \n",
       "\n",
       "                       8             9         ...               4334  \\\n",
       "ctr_DEEx      -38657.800781 -37963.593750      ...      -37493.234375   \n",
       "ctr_upper_IEE -38696.679688 -38067.812500      ...      -37432.890625   \n",
       "ctr_M_G1      -38419.722656 -37928.214844      ...      -37408.769531   \n",
       "ctr_OEE_IEE   -38831.328125 -38218.882812      ...      -37659.273438   \n",
       "ctr_G2_M      -38529.675781 -38074.367188      ...      -37656.128906   \n",
       "ctr_OEE_2     -38712.855469 -38347.917969      ...      -37911.523438   \n",
       "ctr_OSR       -38695.714844 -38299.664062      ...      -37783.640625   \n",
       "ctr_SI        -38743.406250 -38198.730469      ...      -37566.816406   \n",
       "ctr_S         -38600.886719 -37993.128906      ...      -37540.445312   \n",
       "ctr_VEE       -38693.207031 -38509.378906      ...      -38065.285156   \n",
       "ctr_pre_AMB   -38732.742188 -38157.570312      ...      -37423.406250   \n",
       "ctr_OEE_1     -39050.984375 -38500.695312      ...      -38013.066406   \n",
       "ctr_AMB_dist  -39892.593750 -39647.953125      ...      -38947.949219   \n",
       "ctr_AMB_prox  -39482.058594 -39121.781250      ...      -38368.785156   \n",
       "ctr_ISR_SI    -38797.140625 -38369.179688      ...      -37848.394531   \n",
       "\n",
       "                       4335          4336          4337          4338  \\\n",
       "ctr_DEEx      -37796.992188 -38086.953125 -38499.703125 -37928.050781   \n",
       "ctr_upper_IEE -37804.593750 -38370.593750 -38586.027344 -38070.097656   \n",
       "ctr_M_G1      -37672.359375 -38292.304688 -38289.351562 -38093.101562   \n",
       "ctr_OEE_IEE   -37983.074219 -38369.308594 -38710.765625 -37998.824219   \n",
       "ctr_G2_M      -37866.792969 -38396.371094 -38167.925781 -38156.140625   \n",
       "ctr_OEE_2     -38158.085938 -38149.726562 -38825.769531 -37675.179688   \n",
       "ctr_OSR       -38048.109375 -38352.976562 -38723.402344 -37816.804688   \n",
       "ctr_SI        -37883.609375 -38368.410156 -38707.847656 -37883.156250   \n",
       "ctr_S         -37748.078125 -38382.832031 -38282.371094 -38123.441406   \n",
       "ctr_VEE       -38296.089844 -38450.222656 -38989.757812 -37540.753906   \n",
       "ctr_pre_AMB   -37732.039062 -38350.027344 -38639.386719 -37994.761719   \n",
       "ctr_OEE_1     -38286.523438 -38257.335938 -38978.710938 -37981.210938   \n",
       "ctr_AMB_dist  -39086.531250 -39551.707031 -40016.859375 -38868.500000   \n",
       "ctr_AMB_prox  -38536.359375 -39138.968750 -39561.261719 -38505.507812   \n",
       "ctr_ISR_SI    -38103.480469 -38193.117188 -38778.691406 -37830.945312   \n",
       "\n",
       "                       4339          4340          4341          4342  \\\n",
       "ctr_DEEx      -37438.015625 -37693.460938 -38000.359375 -37698.652344   \n",
       "ctr_upper_IEE -37576.816406 -37768.582031 -38050.085938 -37972.800781   \n",
       "ctr_M_G1      -37668.308594 -37772.933594 -37999.277344 -37982.156250   \n",
       "ctr_OEE_IEE   -37353.781250 -37843.414062 -38128.695312 -37844.609375   \n",
       "ctr_G2_M      -37678.015625 -37918.812500 -38147.777344 -38059.562500   \n",
       "ctr_OEE_2     -37156.734375 -37710.210938 -38273.039062 -37458.312500   \n",
       "ctr_OSR       -37350.921875 -37750.824219 -38068.828125 -37815.625000   \n",
       "ctr_SI        -37407.363281 -37587.226562 -37981.140625 -37871.667969   \n",
       "ctr_S         -37640.625000 -37844.835938 -38028.097656 -38010.605469   \n",
       "ctr_VEE       -37356.140625 -37515.898438 -38469.417969 -37611.816406   \n",
       "ctr_pre_AMB   -37521.710938 -37740.382812 -38074.945312 -37939.484375   \n",
       "ctr_OEE_1     -37268.257812 -38017.257812 -38365.375000 -37710.820312   \n",
       "ctr_AMB_dist  -38637.738281 -38798.398438 -39448.601562 -39047.828125   \n",
       "ctr_AMB_prox  -38302.078125 -38372.339844 -38936.535156 -38693.031250   \n",
       "ctr_ISR_SI    -37364.503906 -37787.656250 -38050.359375 -37776.214844   \n",
       "\n",
       "                       4343  \n",
       "ctr_DEEx      -37769.933594  \n",
       "ctr_upper_IEE -37887.433594  \n",
       "ctr_M_G1      -37893.882812  \n",
       "ctr_OEE_IEE   -37867.457031  \n",
       "ctr_G2_M      -37993.644531  \n",
       "ctr_OEE_2     -37666.750000  \n",
       "ctr_OSR       -37756.257812  \n",
       "ctr_SI        -37822.781250  \n",
       "ctr_S         -37947.890625  \n",
       "ctr_VEE       -37702.785156  \n",
       "ctr_pre_AMB   -37864.484375  \n",
       "ctr_OEE_1     -37919.898438  \n",
       "ctr_AMB_dist  -39004.890625  \n",
       "ctr_AMB_prox  -38571.542969  \n",
       "ctr_ISR_SI    -37771.726562  \n",
       "\n",
       "[15 rows x 4344 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# log likelihoods. Can be use directly to find maximum likelihood.\n",
    "bay"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['ctr_M_G1', 'ctr_SI', 'ctr_M_G1', ..., 'ctr_SI', 'ctr_OEE_2',\n",
       "       'ctr_OEE_2'], dtype=object)"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# maximum likelihood for every cell\n",
    "recreated = bay.idxmax().values\n",
    "recreated"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compare to results provided on GEO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "on_geo = adata[toclassify].obs['population'].apply(lambda x:x.replace('inj_','ctr_')).values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 out of 4344 cells have mismatching labels\n"
     ]
    }
   ],
   "source": [
    "mismatches = recreated!=on_geo\n",
    "print(\"%d out of %d cells have mismatching labels\"%(mismatches.sum(),len(mismatches)))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
