{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Annotate class 1 cells by most similar transcriptome among class 2 and 3 populations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os,sys\n",
    "import datetime"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import scanpy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scanpy==1.3.4 anndata==0.6.13 numpy==1.15.4 scipy==1.1.0 pandas==0.23.4 scikit-learn==0.20.1 statsmodels==0.9.0 python-igraph==0.7.1 louvain==0.6.1 \n",
      "Memory usage: current 0.17 GB, difference +0.17 GB\n"
     ]
    }
   ],
   "source": [
    "import scanpy.api as sc\n",
    "sc.logging.print_versions()\n",
    "sc.logging.print_memory_usage()\n",
    "sc.settings.verbosity = 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import my utility functions and import statements from github"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'# This cell is run once to download my custom functions and import statements from github\\n\\n!git clone --depth=1 https://github.com/rapolaszilionis/utility_functions\\n    \\n# github doesn\\'t seem to have an option to download a specific version of the repo from the history.\\n# So I download my utility functions and save the download time by appending it to the directory name.\\n# These utility functions to be shared together with the notebook.\\n\\ntoappend = datetime.datetime.now().strftime(\\'%y%m%d_%Hh%M\\')\\nnewname = \"utility_functions_%s\"%toappend\\nprint(newname)\\n\\n\\n# rename the py file with utility functions\\nos.rename(\"utility_functions\",newname)'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"# This cell is run once to download my custom functions and import statements from github\n",
    "\n",
    "!git clone --depth=1 https://github.com/rapolaszilionis/utility_functions\n",
    "    \n",
    "# github doesn't seem to have an option to download a specific version of the repo from the history.\n",
    "# So I download my utility functions and save the download time by appending it to the directory name.\n",
    "# These utility functions to be shared together with the notebook.\n",
    "\n",
    "toappend = datetime.datetime.now().strftime('%y%m%d_%Hh%M')\n",
    "newname = \"utility_functions_%s\"%toappend\n",
    "print(newname)\n",
    "\n",
    "\n",
    "# rename the py file with utility functions\n",
    "os.rename(\"utility_functions\",newname)\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "python version: 3.6.7\n"
     ]
    }
   ],
   "source": [
    "# add the utility function folder to PATH\n",
    "sys.path.append(os.path.abspath(\"utility_functions_190425_11h32\"))\n",
    "\n",
    "from rz_import_statements import *\n",
    "import rz_functions as rz\n",
    "import rz_fig_params # this adjust mpl.rcParams, almost nothing to import\n",
    "import rz_utility_spring as srz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load data and place into an annData object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cell_info_8594x25.tsv         injury_barcodes.tsv\r\n",
      "control_barcodes.tsv          injury_genes.tsv\r\n",
      "control_genes.tsv             injury_loom.loom\r\n",
      "control_loom.loom             injury_matrix.mtx\r\n",
      "control_matrix.mtx            raw_counts_8594x27998.mtx\r\n",
      "gene_names_alphabetically.txt\r\n"
     ]
    }
   ],
   "source": [
    "!ls data_from_geo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "obspath = 'data_from_geo/cell_info_8594x25.tsv'\n",
    "countpath = 'data_from_geo/raw_counts_8594x27998.mtx'\n",
    "genepath = 'data_from_geo/gene_names_alphabetically.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(8594, 25)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>condition</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>pass_quality_filters</th>\n",
       "      <th>inj_epithelial</th>\n",
       "      <th>ctr_epithelial</th>\n",
       "      <th>excluded_as_immune_or_mesench</th>\n",
       "      <th>class</th>\n",
       "      <th>population</th>\n",
       "      <th>phase</th>\n",
       "      <th>...</th>\n",
       "      <th>x_control</th>\n",
       "      <th>y_control</th>\n",
       "      <th>x_class3_exploded</th>\n",
       "      <th>y_class3_exploded</th>\n",
       "      <th>x_control_injured</th>\n",
       "      <th>y_control_injured</th>\n",
       "      <th>x_class1_CTR_cell_cyc_removed</th>\n",
       "      <th>y_class1_CTR_cell_cyc_removed</th>\n",
       "      <th>x_class1_INJ_cell_cyc_removed</th>\n",
       "      <th>y_class1_INJ_cell_cyc_removed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACCTGAGTGCTGCC-1</td>\n",
       "      <td>control</td>\n",
       "      <td>2787</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>class3</td>\n",
       "      <td>ctr_DEEx</td>\n",
       "      <td>G1</td>\n",
       "      <td>...</td>\n",
       "      <td>783.705500</td>\n",
       "      <td>-377.074709</td>\n",
       "      <td>1082.172898</td>\n",
       "      <td>-266.303304</td>\n",
       "      <td>709.78614</td>\n",
       "      <td>-325.505019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCTGAGTGGGTTG-1</td>\n",
       "      <td>control</td>\n",
       "      <td>3325</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>class3</td>\n",
       "      <td>ctr_upper_IEE</td>\n",
       "      <td>G1</td>\n",
       "      <td>...</td>\n",
       "      <td>709.612642</td>\n",
       "      <td>-429.409941</td>\n",
       "      <td>854.230866</td>\n",
       "      <td>-454.025809</td>\n",
       "      <td>529.15516</td>\n",
       "      <td>-471.961493</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCTGCAAGTCTAC-1</td>\n",
       "      <td>control</td>\n",
       "      <td>1781</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCTGCAATCTGCA-1</td>\n",
       "      <td>control</td>\n",
       "      <td>3468</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCTGCACGGTGTC-1</td>\n",
       "      <td>control</td>\n",
       "      <td>1745</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              barcode condition  total_counts  pass_quality_filters  \\\n",
       "0  AAACCTGAGTGCTGCC-1   control          2787                  True   \n",
       "1  AAACCTGAGTGGGTTG-1   control          3325                  True   \n",
       "2  AAACCTGCAAGTCTAC-1   control          1781                 False   \n",
       "3  AAACCTGCAATCTGCA-1   control          3468                  True   \n",
       "4  AAACCTGCACGGTGTC-1   control          1745                 False   \n",
       "\n",
       "   inj_epithelial  ctr_epithelial  excluded_as_immune_or_mesench   class  \\\n",
       "0           False            True                          False  class3   \n",
       "1           False            True                          False  class3   \n",
       "2           False           False                          False     NaN   \n",
       "3           False           False                           True     NaN   \n",
       "4           False           False                          False     NaN   \n",
       "\n",
       "      population phase              ...                 x_control   y_control  \\\n",
       "0       ctr_DEEx    G1              ...                783.705500 -377.074709   \n",
       "1  ctr_upper_IEE    G1              ...                709.612642 -429.409941   \n",
       "2            NaN   NaN              ...                       NaN         NaN   \n",
       "3            NaN   NaN              ...                       NaN         NaN   \n",
       "4            NaN   NaN              ...                       NaN         NaN   \n",
       "\n",
       "  x_class3_exploded  y_class3_exploded x_control_injured  y_control_injured  \\\n",
       "0       1082.172898        -266.303304         709.78614        -325.505019   \n",
       "1        854.230866        -454.025809         529.15516        -471.961493   \n",
       "2               NaN                NaN               NaN                NaN   \n",
       "3               NaN                NaN               NaN                NaN   \n",
       "4               NaN                NaN               NaN                NaN   \n",
       "\n",
       "   x_class1_CTR_cell_cyc_removed  y_class1_CTR_cell_cyc_removed  \\\n",
       "0                            NaN                            NaN   \n",
       "1                            NaN                            NaN   \n",
       "2                            NaN                            NaN   \n",
       "3                            NaN                            NaN   \n",
       "4                            NaN                            NaN   \n",
       "\n",
       "   x_class1_INJ_cell_cyc_removed  y_class1_INJ_cell_cyc_removed  \n",
       "0                            NaN                            NaN  \n",
       "1                            NaN                            NaN  \n",
       "2                            NaN                            NaN  \n",
       "3                            NaN                            NaN  \n",
       "4                            NaN                            NaN  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load cell info\n",
    "obs = pd.read_csv(obspath,sep='\\t',index_col=0)\n",
    "print(obs.shape)\n",
    "obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(8594, 27998)\n"
     ]
    }
   ],
   "source": [
    "# load counts\n",
    "adata = sc.read_mtx(countpath)\n",
    "print(adata.shape)\n",
    "\n",
    "# add genes (annotation of variables)\n",
    "adata.var['genes'] = np.loadtxt(genepath,dtype=str)\n",
    "\n",
    "# make sure var names are genes\n",
    "adata.var_names = adata.var['genes'].values\n",
    "\n",
    "# add obs (annotation of observations)\n",
    "adata.obs = obs\n",
    "\n",
    "# make sure index is unique AND a string\n",
    "adata.obs_names_make_unique()\n",
    "adata.obs_names = adata.obs_names.astype(str)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Scale (normalize) data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[2787.]\n",
      " [3325.]\n",
      " [1781.]\n",
      " [3468.]\n",
      " [1745.]]\n",
      "\n",
      "[[10000.   ]\n",
      " [10000.   ]\n",
      " [10000.001]\n",
      " [10000.   ]\n",
      " [ 9999.999]]\n"
     ]
    }
   ],
   "source": [
    "# turn into counts per 10k\n",
    "print(adata.X[:5,:].sum(axis=1))\n",
    "print()\n",
    "sc.pp.normalize_per_cell(adata, counts_per_cell_after=1e4)\n",
    "print(adata.X[:5,:].sum(axis=1))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get centroids (reference transcriptomes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function centroids in module rz_functions:\n",
      "\n",
      "centroids(label, adata, E=None, gene_list=None)\n",
      "    Calculate average gene expression level per cell label (e.g. cluster).\n",
      "    input:\n",
      "        - label: name of column that stores the label of interest in adata.obs\n",
      "        - adata: AnnData object OR a cell x feature pandas dataframe with label as one of the columns\n",
      "        - E and gene_list: optional and only used when adata is not an AnnData object. In that case\n",
      "        the cells x genes sparse expression matrix E and the gene_list must be specified\n",
      "        \n",
      "    returns:\n",
      "        pandas dataframe, centroids x genes\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(rz.centroids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "label = 'population'\n",
    "cell_mask = adata.obs['ctr_epithelial'].astype(bool).values\n",
    "centroids = rz.centroids(label,adata[cell_mask])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0610007P14Rik</th>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <th>0610009L18Rik</th>\n",
       "      <th>0610009O20Rik</th>\n",
       "      <th>0610010F05Rik</th>\n",
       "      <th>0610010K14Rik</th>\n",
       "      <th>0610011F06Rik</th>\n",
       "      <th>0610012D04Rik</th>\n",
       "      <th>0610012G03Rik</th>\n",
       "      <th>0610025J13Rik</th>\n",
       "      <th>...</th>\n",
       "      <th>mt-Co2</th>\n",
       "      <th>mt-Co3</th>\n",
       "      <th>mt-Cytb</th>\n",
       "      <th>mt-Nd1</th>\n",
       "      <th>mt-Nd2</th>\n",
       "      <th>mt-Nd3</th>\n",
       "      <th>mt-Nd4</th>\n",
       "      <th>mt-Nd4l</th>\n",
       "      <th>mt-Nd5</th>\n",
       "      <th>mt-Nd6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ctr_DEEx</th>\n",
       "      <td>1.391852</td>\n",
       "      <td>0.879937</td>\n",
       "      <td>0.324726</td>\n",
       "      <td>0.209586</td>\n",
       "      <td>0.086968</td>\n",
       "      <td>0.329960</td>\n",
       "      <td>0.954454</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.200782</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>14.079573</td>\n",
       "      <td>83.904465</td>\n",
       "      <td>56.074482</td>\n",
       "      <td>23.852005</td>\n",
       "      <td>3.198850</td>\n",
       "      <td>1.030915</td>\n",
       "      <td>18.962263</td>\n",
       "      <td>0.987729</td>\n",
       "      <td>1.261443</td>\n",
       "      <td>0.275973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctr_upper_IEE</th>\n",
       "      <td>1.300044</td>\n",
       "      <td>0.695099</td>\n",
       "      <td>0.314438</td>\n",
       "      <td>0.287892</td>\n",
       "      <td>0.184926</td>\n",
       "      <td>0.397997</td>\n",
       "      <td>0.773451</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.998683</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>15.316485</td>\n",
       "      <td>84.531807</td>\n",
       "      <td>54.308296</td>\n",
       "      <td>24.424025</td>\n",
       "      <td>3.601944</td>\n",
       "      <td>1.055761</td>\n",
       "      <td>18.239050</td>\n",
       "      <td>0.895775</td>\n",
       "      <td>0.950987</td>\n",
       "      <td>0.232046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctr_M_G1</th>\n",
       "      <td>1.303535</td>\n",
       "      <td>0.552314</td>\n",
       "      <td>0.065731</td>\n",
       "      <td>0.330279</td>\n",
       "      <td>0.126810</td>\n",
       "      <td>0.566542</td>\n",
       "      <td>0.606099</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.455639</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>12.348806</td>\n",
       "      <td>70.209267</td>\n",
       "      <td>47.538788</td>\n",
       "      <td>23.879992</td>\n",
       "      <td>2.751078</td>\n",
       "      <td>0.998057</td>\n",
       "      <td>15.154018</td>\n",
       "      <td>0.907756</td>\n",
       "      <td>0.761977</td>\n",
       "      <td>0.202841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctr_OEE_IEE</th>\n",
       "      <td>1.737094</td>\n",
       "      <td>0.544913</td>\n",
       "      <td>0.267885</td>\n",
       "      <td>0.461709</td>\n",
       "      <td>0.105900</td>\n",
       "      <td>0.447238</td>\n",
       "      <td>0.927448</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.360467</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>13.681261</td>\n",
       "      <td>85.969254</td>\n",
       "      <td>55.733368</td>\n",
       "      <td>22.308928</td>\n",
       "      <td>4.568748</td>\n",
       "      <td>0.957020</td>\n",
       "      <td>19.031935</td>\n",
       "      <td>0.749691</td>\n",
       "      <td>1.071926</td>\n",
       "      <td>0.076110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctr_G2_M</th>\n",
       "      <td>2.209283</td>\n",
       "      <td>0.774067</td>\n",
       "      <td>0.063674</td>\n",
       "      <td>0.473809</td>\n",
       "      <td>0.182266</td>\n",
       "      <td>0.590541</td>\n",
       "      <td>0.573025</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.316742</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>12.441298</td>\n",
       "      <td>70.319290</td>\n",
       "      <td>45.029736</td>\n",
       "      <td>23.863428</td>\n",
       "      <td>3.214058</td>\n",
       "      <td>0.953022</td>\n",
       "      <td>15.301864</td>\n",
       "      <td>0.760218</td>\n",
       "      <td>1.103230</td>\n",
       "      <td>0.170797</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 27998 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               0610007P14Rik  0610009B22Rik  0610009L18Rik  0610009O20Rik  \\\n",
       "ctr_DEEx            1.391852       0.879937       0.324726       0.209586   \n",
       "ctr_upper_IEE       1.300044       0.695099       0.314438       0.287892   \n",
       "ctr_M_G1            1.303535       0.552314       0.065731       0.330279   \n",
       "ctr_OEE_IEE         1.737094       0.544913       0.267885       0.461709   \n",
       "ctr_G2_M            2.209283       0.774067       0.063674       0.473809   \n",
       "\n",
       "               0610010F05Rik  0610010K14Rik  0610011F06Rik  0610012D04Rik  \\\n",
       "ctr_DEEx            0.086968       0.329960       0.954454            0.0   \n",
       "ctr_upper_IEE       0.184926       0.397997       0.773451            0.0   \n",
       "ctr_M_G1            0.126810       0.566542       0.606099            0.0   \n",
       "ctr_OEE_IEE         0.105900       0.447238       0.927448            0.0   \n",
       "ctr_G2_M            0.182266       0.590541       0.573025            0.0   \n",
       "\n",
       "               0610012G03Rik  0610025J13Rik    ...        mt-Co2     mt-Co3  \\\n",
       "ctr_DEEx            1.200782            0.0    ...     14.079573  83.904465   \n",
       "ctr_upper_IEE       0.998683            0.0    ...     15.316485  84.531807   \n",
       "ctr_M_G1            1.455639            0.0    ...     12.348806  70.209267   \n",
       "ctr_OEE_IEE         1.360467            0.0    ...     13.681261  85.969254   \n",
       "ctr_G2_M            1.316742            0.0    ...     12.441298  70.319290   \n",
       "\n",
       "                 mt-Cytb     mt-Nd1    mt-Nd2    mt-Nd3     mt-Nd4   mt-Nd4l  \\\n",
       "ctr_DEEx       56.074482  23.852005  3.198850  1.030915  18.962263  0.987729   \n",
       "ctr_upper_IEE  54.308296  24.424025  3.601944  1.055761  18.239050  0.895775   \n",
       "ctr_M_G1       47.538788  23.879992  2.751078  0.998057  15.154018  0.907756   \n",
       "ctr_OEE_IEE    55.733368  22.308928  4.568748  0.957020  19.031935  0.749691   \n",
       "ctr_G2_M       45.029736  23.863428  3.214058  0.953022  15.301864  0.760218   \n",
       "\n",
       "                 mt-Nd5    mt-Nd6  \n",
       "ctr_DEEx       1.261443  0.275973  \n",
       "ctr_upper_IEE  0.950987  0.232046  \n",
       "ctr_M_G1       0.761977  0.202841  \n",
       "ctr_OEE_IEE    1.071926  0.076110  \n",
       "ctr_G2_M       1.103230  0.170797  \n",
       "\n",
       "[5 rows x 27998 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "centroids.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run classifier with 100 most enriched genes per class2,3 population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3344 8594\n",
      "['ctr_DEEx', 'ctr_upper_IEE', 'ctr_OEE_IEE', 'ctr_OEE_2', 'ctr_OSR', 'ctr_SI', 'ctr_VEE', 'ctr_pre_AMB', 'ctr_OEE_1', 'ctr_AMB_dist', 'ctr_AMB_prox', 'ctr_ISR_SI']\n"
     ]
    }
   ],
   "source": [
    "# select cells to classify\n",
    "toclassify = adata.obs['class'].isin(['class1'])&\\\n",
    "            (adata.obs['ctr_epithelial'].astype(bool)|adata.obs['inj_epithelial'].astype(bool))\n",
    "toclassify = toclassify.values\n",
    "print(toclassify.sum(),len(toclassify))\n",
    "E = adata[toclassify].X\n",
    "gene_list = adata.var_names\n",
    "\n",
    "# reference transcriptomes to classify by\n",
    "category_labels = [i for i in centroids.index if i not in adata[toclassify].obs['population'].unique()]\n",
    "categories = centroids.T[category_labels]\n",
    "print(category_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ctr_DEEx</th>\n",
       "      <th>ctr_DEEx_FC</th>\n",
       "      <th>ctr_upper_IEE</th>\n",
       "      <th>ctr_upper_IEE_FC</th>\n",
       "      <th>ctr_M_G1</th>\n",
       "      <th>ctr_M_G1_FC</th>\n",
       "      <th>ctr_OEE_IEE</th>\n",
       "      <th>ctr_OEE_IEE_FC</th>\n",
       "      <th>ctr_G2_M</th>\n",
       "      <th>ctr_G2_M_FC</th>\n",
       "      <th>...</th>\n",
       "      <th>ctr_pre_AMB</th>\n",
       "      <th>ctr_pre_AMB_FC</th>\n",
       "      <th>ctr_OEE_1</th>\n",
       "      <th>ctr_OEE_1_FC</th>\n",
       "      <th>ctr_AMB_dist</th>\n",
       "      <th>ctr_AMB_dist_FC</th>\n",
       "      <th>ctr_AMB_prox</th>\n",
       "      <th>ctr_AMB_prox_FC</th>\n",
       "      <th>ctr_ISR_SI</th>\n",
       "      <th>ctr_ISR_SI_FC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Dcn</td>\n",
       "      <td>3.614268</td>\n",
       "      <td>Rnaset2a</td>\n",
       "      <td>2.148060</td>\n",
       "      <td>Cenpa</td>\n",
       "      <td>3.363334</td>\n",
       "      <td>Cldn10</td>\n",
       "      <td>5.468689</td>\n",
       "      <td>Ube2c</td>\n",
       "      <td>7.981552</td>\n",
       "      <td>...</td>\n",
       "      <td>Nudt4</td>\n",
       "      <td>2.511703</td>\n",
       "      <td>Grp</td>\n",
       "      <td>7.475961</td>\n",
       "      <td>Ambn</td>\n",
       "      <td>79.594589</td>\n",
       "      <td>Amelx</td>\n",
       "      <td>40.556911</td>\n",
       "      <td>Krt17</td>\n",
       "      <td>5.546733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Aldh1a3</td>\n",
       "      <td>3.171303</td>\n",
       "      <td>Phlda1</td>\n",
       "      <td>1.994479</td>\n",
       "      <td>Hist1h2bc</td>\n",
       "      <td>3.181940</td>\n",
       "      <td>Pthlh</td>\n",
       "      <td>4.153130</td>\n",
       "      <td>Top2a</td>\n",
       "      <td>6.105304</td>\n",
       "      <td>...</td>\n",
       "      <td>Iqgap2</td>\n",
       "      <td>2.222442</td>\n",
       "      <td>Nppc</td>\n",
       "      <td>6.380324</td>\n",
       "      <td>Enam</td>\n",
       "      <td>44.962585</td>\n",
       "      <td>Calb1</td>\n",
       "      <td>37.223640</td>\n",
       "      <td>Tacstd2</td>\n",
       "      <td>4.155315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Sfrp5</td>\n",
       "      <td>2.575278</td>\n",
       "      <td>Ifitm3</td>\n",
       "      <td>1.992567</td>\n",
       "      <td>Ccnb2</td>\n",
       "      <td>2.390234</td>\n",
       "      <td>Grp</td>\n",
       "      <td>3.896335</td>\n",
       "      <td>Kpna2</td>\n",
       "      <td>5.745326</td>\n",
       "      <td>...</td>\n",
       "      <td>Satb1</td>\n",
       "      <td>2.204441</td>\n",
       "      <td>Fosb</td>\n",
       "      <td>6.052104</td>\n",
       "      <td>Amelx</td>\n",
       "      <td>41.222912</td>\n",
       "      <td>Enam</td>\n",
       "      <td>35.814861</td>\n",
       "      <td>Tagln</td>\n",
       "      <td>3.631651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Ccnd2</td>\n",
       "      <td>2.470626</td>\n",
       "      <td>Sox4</td>\n",
       "      <td>1.895130</td>\n",
       "      <td>Hmgb2</td>\n",
       "      <td>2.380920</td>\n",
       "      <td>Crabp1</td>\n",
       "      <td>3.029581</td>\n",
       "      <td>Nusap1</td>\n",
       "      <td>5.553847</td>\n",
       "      <td>...</td>\n",
       "      <td>Col27a1</td>\n",
       "      <td>2.184898</td>\n",
       "      <td>Dcn</td>\n",
       "      <td>4.905131</td>\n",
       "      <td>Clu</td>\n",
       "      <td>39.412006</td>\n",
       "      <td>Clu</td>\n",
       "      <td>23.112156</td>\n",
       "      <td>Igfbp3</td>\n",
       "      <td>3.444902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>En1</td>\n",
       "      <td>2.386213</td>\n",
       "      <td>Fam19a4</td>\n",
       "      <td>1.858609</td>\n",
       "      <td>Lockd</td>\n",
       "      <td>2.336322</td>\n",
       "      <td>Shisa2</td>\n",
       "      <td>2.936753</td>\n",
       "      <td>Ccnb1</td>\n",
       "      <td>5.493615</td>\n",
       "      <td>...</td>\n",
       "      <td>Vwde</td>\n",
       "      <td>2.173148</td>\n",
       "      <td>Tgfb2</td>\n",
       "      <td>4.904398</td>\n",
       "      <td>Mmp20</td>\n",
       "      <td>14.765478</td>\n",
       "      <td>Plod2</td>\n",
       "      <td>22.421888</td>\n",
       "      <td>Slc39a10</td>\n",
       "      <td>3.411926</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  ctr_DEEx  ctr_DEEx_FC ctr_upper_IEE  ctr_upper_IEE_FC   ctr_M_G1  \\\n",
       "0      Dcn     3.614268      Rnaset2a          2.148060      Cenpa   \n",
       "1  Aldh1a3     3.171303        Phlda1          1.994479  Hist1h2bc   \n",
       "2    Sfrp5     2.575278        Ifitm3          1.992567      Ccnb2   \n",
       "3    Ccnd2     2.470626          Sox4          1.895130      Hmgb2   \n",
       "4      En1     2.386213       Fam19a4          1.858609      Lockd   \n",
       "\n",
       "   ctr_M_G1_FC ctr_OEE_IEE  ctr_OEE_IEE_FC ctr_G2_M  ctr_G2_M_FC  \\\n",
       "0     3.363334      Cldn10        5.468689    Ube2c     7.981552   \n",
       "1     3.181940       Pthlh        4.153130    Top2a     6.105304   \n",
       "2     2.390234         Grp        3.896335    Kpna2     5.745326   \n",
       "3     2.380920      Crabp1        3.029581   Nusap1     5.553847   \n",
       "4     2.336322      Shisa2        2.936753    Ccnb1     5.493615   \n",
       "\n",
       "       ...       ctr_pre_AMB  ctr_pre_AMB_FC ctr_OEE_1  ctr_OEE_1_FC  \\\n",
       "0      ...             Nudt4        2.511703       Grp      7.475961   \n",
       "1      ...            Iqgap2        2.222442      Nppc      6.380324   \n",
       "2      ...             Satb1        2.204441      Fosb      6.052104   \n",
       "3      ...           Col27a1        2.184898       Dcn      4.905131   \n",
       "4      ...              Vwde        2.173148     Tgfb2      4.904398   \n",
       "\n",
       "  ctr_AMB_dist  ctr_AMB_dist_FC ctr_AMB_prox  ctr_AMB_prox_FC ctr_ISR_SI  \\\n",
       "0         Ambn        79.594589        Amelx        40.556911      Krt17   \n",
       "1         Enam        44.962585        Calb1        37.223640    Tacstd2   \n",
       "2        Amelx        41.222912         Enam        35.814861      Tagln   \n",
       "3          Clu        39.412006          Clu        23.112156     Igfbp3   \n",
       "4        Mmp20        14.765478        Plod2        22.421888   Slc39a10   \n",
       "\n",
       "   ctr_ISR_SI_FC  \n",
       "0       5.546733  \n",
       "1       4.155315  \n",
       "2       3.631651  \n",
       "3       3.444902  \n",
       "4       3.411926  \n",
       "\n",
       "[5 rows x 30 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load table with top 100 most enriched genes per population\n",
    "top100 = pd.read_excel('supplementary_tables_Sharir_et_al_2019/Table_S1.xlsx',header=1)\n",
    "\n",
    "top100.columns = [i.replace('/','_').replace(' ','_').replace('-','_') for i in top100.columns]\n",
    "top100.columns = ['ctr_'+i for i in top100.columns]\n",
    "top100.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1200\n",
      "800\n",
      "790\n"
     ]
    }
   ],
   "source": [
    "# enriched genes in clusters outside of cycling cells\n",
    "enriched_genes = top100[category_labels].values.flatten()\n",
    "labels_to_avoid = [i for i in centroids.index if i not in category_labels]\n",
    "genes_to_avoid = top100[labels_to_avoid].values.flatten()\n",
    "print(len(enriched_genes))\n",
    "enriched_genes = list(set(enriched_genes))\n",
    "print(len(enriched_genes))\n",
    "# exclude occasional cell cycle gene\n",
    "enriched_genes = [i for i in enriched_genes if i not in genes_to_avoid]\n",
    "print(len(enriched_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select genes to use, prepare boolean mask.\n",
    "genes_to_use = enriched_genes\n",
    "gene_mask = np.in1d(gene_list,genes_to_use)\n",
    "\n",
    "pseudo = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.00 min.\n",
      "cells from 0 to 1000 done\n",
      "0.01 min.\n",
      "cells from 1000 to 2000 done\n",
      "0.01 min.\n",
      "cells from 2000 to 3000 done\n",
      "0.01 min.\n",
      "cells from 3000 to 3344 done\n"
     ]
    }
   ],
   "source": [
    "start = time.time()\n",
    "bays = []\n",
    "i = 0\n",
    "interval = 1000\n",
    "for j in range(interval,E.shape[0]+interval,interval):\n",
    "    j = min(j,E.shape[0])\n",
    "    stepsize = j-i\n",
    "    tmp_dense = pd.DataFrame(E.T[gene_mask][:,i:j].todense())\n",
    "    tmp_dense.index = np.array(gene_list)[gene_mask]\n",
    "    bay = rz.bayesian_classifier(tmp_dense,categories.loc[tmp_dense.index]+pseudo)\n",
    "    bays.append(bay)\n",
    "    i = j\n",
    "    print('%.2f min.'%((time.time()-start)/60.))\n",
    "    print('cells from %d to %d done'%(i-stepsize,min(i,E.shape[0])))\n",
    "    \n",
    "bay = pd.concat(bays,axis=1)\n",
    "bay.columns = np.arange(bay.shape[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>3334</th>\n",
       "      <th>3335</th>\n",
       "      <th>3336</th>\n",
       "      <th>3337</th>\n",
       "      <th>3338</th>\n",
       "      <th>3339</th>\n",
       "      <th>3340</th>\n",
       "      <th>3341</th>\n",
       "      <th>3342</th>\n",
       "      <th>3343</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ctr_DEEx</th>\n",
       "      <td>-4031.168457</td>\n",
       "      <td>-3819.042480</td>\n",
       "      <td>-3348.066895</td>\n",
       "      <td>-4054.513184</td>\n",
       "      <td>-3619.425293</td>\n",
       "      <td>-5047.666016</td>\n",
       "      <td>-3317.287109</td>\n",
       "      <td>-4297.836426</td>\n",
       "      <td>-3743.825195</td>\n",
       "      <td>-3645.994629</td>\n",
       "      <td>...</td>\n",
       "      <td>-3718.858398</td>\n",
       "      <td>-4774.810547</td>\n",
       "      <td>-3405.502441</td>\n",
       "      <td>-3209.570312</td>\n",
       "      <td>-3864.764893</td>\n",
       "      <td>-3724.276855</td>\n",
       "      <td>-4436.709473</td>\n",
       "      <td>-4342.840820</td>\n",
       "      <td>-4401.244629</td>\n",
       "      <td>-4121.519531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctr_upper_IEE</th>\n",
       "      <td>-4018.796631</td>\n",
       "      <td>-3802.652832</td>\n",
       "      <td>-3302.541504</td>\n",
       "      <td>-4000.065918</td>\n",
       "      <td>-3571.762207</td>\n",
       "      <td>-5118.182617</td>\n",
       "      <td>-3299.607666</td>\n",
       "      <td>-4274.031250</td>\n",
       "      <td>-3733.900879</td>\n",
       "      <td>-3626.227539</td>\n",
       "      <td>...</td>\n",
       "      <td>-3714.743164</td>\n",
       "      <td>-4835.724609</td>\n",
       "      <td>-3373.500000</td>\n",
       "      <td>-3220.977295</td>\n",
       "      <td>-3836.332520</td>\n",
       "      <td>-3723.338379</td>\n",
       "      <td>-4413.879883</td>\n",
       "      <td>-4315.059082</td>\n",
       "      <td>-4414.622070</td>\n",
       "      <td>-4197.270508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctr_OEE_IEE</th>\n",
       "      <td>-4082.834961</td>\n",
       "      <td>-3848.474365</td>\n",
       "      <td>-3354.785156</td>\n",
       "      <td>-3944.142578</td>\n",
       "      <td>-3615.201172</td>\n",
       "      <td>-5163.550781</td>\n",
       "      <td>-3331.181641</td>\n",
       "      <td>-4298.198242</td>\n",
       "      <td>-3771.521729</td>\n",
       "      <td>-3695.071777</td>\n",
       "      <td>...</td>\n",
       "      <td>-3721.506104</td>\n",
       "      <td>-4784.722656</td>\n",
       "      <td>-3469.545166</td>\n",
       "      <td>-3257.449463</td>\n",
       "      <td>-3906.003662</td>\n",
       "      <td>-3803.436523</td>\n",
       "      <td>-4337.062500</td>\n",
       "      <td>-4443.498047</td>\n",
       "      <td>-4507.806152</td>\n",
       "      <td>-4243.020020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctr_OEE_2</th>\n",
       "      <td>-4123.669922</td>\n",
       "      <td>-3926.993408</td>\n",
       "      <td>-3445.233398</td>\n",
       "      <td>-4103.208984</td>\n",
       "      <td>-3727.003418</td>\n",
       "      <td>-5001.466797</td>\n",
       "      <td>-3432.708984</td>\n",
       "      <td>-4385.477539</td>\n",
       "      <td>-3836.349854</td>\n",
       "      <td>-3766.377686</td>\n",
       "      <td>...</td>\n",
       "      <td>-3784.227539</td>\n",
       "      <td>-4701.400391</td>\n",
       "      <td>-3548.579102</td>\n",
       "      <td>-3312.280762</td>\n",
       "      <td>-3954.685791</td>\n",
       "      <td>-3856.008545</td>\n",
       "      <td>-4407.008789</td>\n",
       "      <td>-4508.538086</td>\n",
       "      <td>-4547.631836</td>\n",
       "      <td>-4244.314453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctr_OSR</th>\n",
       "      <td>-4073.936035</td>\n",
       "      <td>-3880.292969</td>\n",
       "      <td>-3368.363525</td>\n",
       "      <td>-4003.779785</td>\n",
       "      <td>-3695.619385</td>\n",
       "      <td>-5140.972168</td>\n",
       "      <td>-3396.118164</td>\n",
       "      <td>-4286.783203</td>\n",
       "      <td>-3801.904053</td>\n",
       "      <td>-3734.900635</td>\n",
       "      <td>...</td>\n",
       "      <td>-3771.967529</td>\n",
       "      <td>-4659.078613</td>\n",
       "      <td>-3516.998535</td>\n",
       "      <td>-3315.391846</td>\n",
       "      <td>-3826.553467</td>\n",
       "      <td>-3841.211670</td>\n",
       "      <td>-4255.573242</td>\n",
       "      <td>-4489.912598</td>\n",
       "      <td>-4537.487305</td>\n",
       "      <td>-4261.460938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctr_SI</th>\n",
       "      <td>-4040.829102</td>\n",
       "      <td>-3855.736328</td>\n",
       "      <td>-3334.332520</td>\n",
       "      <td>-4002.202148</td>\n",
       "      <td>-3623.750000</td>\n",
       "      <td>-5152.177734</td>\n",
       "      <td>-3362.491699</td>\n",
       "      <td>-4231.451172</td>\n",
       "      <td>-3749.849609</td>\n",
       "      <td>-3664.914307</td>\n",
       "      <td>...</td>\n",
       "      <td>-3748.192139</td>\n",
       "      <td>-4702.145996</td>\n",
       "      <td>-3454.336426</td>\n",
       "      <td>-3279.175781</td>\n",
       "      <td>-3783.816895</td>\n",
       "      <td>-3782.928955</td>\n",
       "      <td>-4315.301758</td>\n",
       "      <td>-4397.071289</td>\n",
       "      <td>-4464.516113</td>\n",
       "      <td>-4236.380859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctr_VEE</th>\n",
       "      <td>-4140.198242</td>\n",
       "      <td>-3991.459229</td>\n",
       "      <td>-3465.568848</td>\n",
       "      <td>-4114.451172</td>\n",
       "      <td>-3776.950684</td>\n",
       "      <td>-5093.940918</td>\n",
       "      <td>-3489.823242</td>\n",
       "      <td>-4465.648438</td>\n",
       "      <td>-3860.351074</td>\n",
       "      <td>-3780.692383</td>\n",
       "      <td>...</td>\n",
       "      <td>-3832.022949</td>\n",
       "      <td>-4754.228516</td>\n",
       "      <td>-3553.413574</td>\n",
       "      <td>-3355.746582</td>\n",
       "      <td>-3998.401367</td>\n",
       "      <td>-3882.810303</td>\n",
       "      <td>-4483.126953</td>\n",
       "      <td>-4555.071289</td>\n",
       "      <td>-4568.892578</td>\n",
       "      <td>-4286.059570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctr_pre_AMB</th>\n",
       "      <td>-3971.533691</td>\n",
       "      <td>-3822.525391</td>\n",
       "      <td>-3348.704102</td>\n",
       "      <td>-4074.522949</td>\n",
       "      <td>-3575.081055</td>\n",
       "      <td>-5142.391602</td>\n",
       "      <td>-3328.959717</td>\n",
       "      <td>-4333.919922</td>\n",
       "      <td>-3689.512207</td>\n",
       "      <td>-3595.837646</td>\n",
       "      <td>...</td>\n",
       "      <td>-3756.101562</td>\n",
       "      <td>-4803.732422</td>\n",
       "      <td>-3413.587402</td>\n",
       "      <td>-3270.857422</td>\n",
       "      <td>-3883.302246</td>\n",
       "      <td>-3745.017822</td>\n",
       "      <td>-4435.564453</td>\n",
       "      <td>-4301.487305</td>\n",
       "      <td>-4366.380371</td>\n",
       "      <td>-4199.173828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctr_OEE_1</th>\n",
       "      <td>-4168.995117</td>\n",
       "      <td>-3952.783447</td>\n",
       "      <td>-3490.737549</td>\n",
       "      <td>-4115.011230</td>\n",
       "      <td>-3746.059326</td>\n",
       "      <td>-5191.430664</td>\n",
       "      <td>-3425.160156</td>\n",
       "      <td>-4422.537109</td>\n",
       "      <td>-3849.865967</td>\n",
       "      <td>-3806.821045</td>\n",
       "      <td>...</td>\n",
       "      <td>-3823.564941</td>\n",
       "      <td>-4748.683594</td>\n",
       "      <td>-3615.694824</td>\n",
       "      <td>-3367.099365</td>\n",
       "      <td>-4038.809570</td>\n",
       "      <td>-3922.755127</td>\n",
       "      <td>-4444.573242</td>\n",
       "      <td>-4587.330078</td>\n",
       "      <td>-4635.699219</td>\n",
       "      <td>-4315.931152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctr_AMB_dist</th>\n",
       "      <td>-4517.004883</td>\n",
       "      <td>-4381.872070</td>\n",
       "      <td>-3891.899414</td>\n",
       "      <td>-4596.195801</td>\n",
       "      <td>-4184.358887</td>\n",
       "      <td>-5787.214844</td>\n",
       "      <td>-3891.475586</td>\n",
       "      <td>-4949.999023</td>\n",
       "      <td>-4254.697266</td>\n",
       "      <td>-4150.666504</td>\n",
       "      <td>...</td>\n",
       "      <td>-4347.254883</td>\n",
       "      <td>-5462.891602</td>\n",
       "      <td>-3977.195801</td>\n",
       "      <td>-3789.163086</td>\n",
       "      <td>-4452.683594</td>\n",
       "      <td>-4317.268066</td>\n",
       "      <td>-5040.771484</td>\n",
       "      <td>-4979.303223</td>\n",
       "      <td>-5000.744629</td>\n",
       "      <td>-4838.551270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctr_AMB_prox</th>\n",
       "      <td>-4342.791016</td>\n",
       "      <td>-4232.943359</td>\n",
       "      <td>-3738.455566</td>\n",
       "      <td>-4463.638672</td>\n",
       "      <td>-4025.751953</td>\n",
       "      <td>-5646.947754</td>\n",
       "      <td>-3745.275391</td>\n",
       "      <td>-4811.572754</td>\n",
       "      <td>-4079.221680</td>\n",
       "      <td>-3988.664795</td>\n",
       "      <td>...</td>\n",
       "      <td>-4210.827148</td>\n",
       "      <td>-5303.000488</td>\n",
       "      <td>-3820.992676</td>\n",
       "      <td>-3653.866211</td>\n",
       "      <td>-4309.155273</td>\n",
       "      <td>-4161.617676</td>\n",
       "      <td>-4914.133789</td>\n",
       "      <td>-4790.520996</td>\n",
       "      <td>-4803.109375</td>\n",
       "      <td>-4669.141113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ctr_ISR_SI</th>\n",
       "      <td>-4100.297852</td>\n",
       "      <td>-3903.675293</td>\n",
       "      <td>-3406.209961</td>\n",
       "      <td>-4061.554199</td>\n",
       "      <td>-3731.051270</td>\n",
       "      <td>-5124.352051</td>\n",
       "      <td>-3417.758301</td>\n",
       "      <td>-4320.359375</td>\n",
       "      <td>-3805.827881</td>\n",
       "      <td>-3747.845215</td>\n",
       "      <td>...</td>\n",
       "      <td>-3810.211914</td>\n",
       "      <td>-4591.441406</td>\n",
       "      <td>-3556.550537</td>\n",
       "      <td>-3340.044434</td>\n",
       "      <td>-3847.442627</td>\n",
       "      <td>-3855.660156</td>\n",
       "      <td>-4313.973633</td>\n",
       "      <td>-4508.205078</td>\n",
       "      <td>-4543.085449</td>\n",
       "      <td>-4237.670410</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12 rows × 3344 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      0            1            2            3     \\\n",
       "ctr_DEEx      -4031.168457 -3819.042480 -3348.066895 -4054.513184   \n",
       "ctr_upper_IEE -4018.796631 -3802.652832 -3302.541504 -4000.065918   \n",
       "ctr_OEE_IEE   -4082.834961 -3848.474365 -3354.785156 -3944.142578   \n",
       "ctr_OEE_2     -4123.669922 -3926.993408 -3445.233398 -4103.208984   \n",
       "ctr_OSR       -4073.936035 -3880.292969 -3368.363525 -4003.779785   \n",
       "ctr_SI        -4040.829102 -3855.736328 -3334.332520 -4002.202148   \n",
       "ctr_VEE       -4140.198242 -3991.459229 -3465.568848 -4114.451172   \n",
       "ctr_pre_AMB   -3971.533691 -3822.525391 -3348.704102 -4074.522949   \n",
       "ctr_OEE_1     -4168.995117 -3952.783447 -3490.737549 -4115.011230   \n",
       "ctr_AMB_dist  -4517.004883 -4381.872070 -3891.899414 -4596.195801   \n",
       "ctr_AMB_prox  -4342.791016 -4232.943359 -3738.455566 -4463.638672   \n",
       "ctr_ISR_SI    -4100.297852 -3903.675293 -3406.209961 -4061.554199   \n",
       "\n",
       "                      4            5            6            7     \\\n",
       "ctr_DEEx      -3619.425293 -5047.666016 -3317.287109 -4297.836426   \n",
       "ctr_upper_IEE -3571.762207 -5118.182617 -3299.607666 -4274.031250   \n",
       "ctr_OEE_IEE   -3615.201172 -5163.550781 -3331.181641 -4298.198242   \n",
       "ctr_OEE_2     -3727.003418 -5001.466797 -3432.708984 -4385.477539   \n",
       "ctr_OSR       -3695.619385 -5140.972168 -3396.118164 -4286.783203   \n",
       "ctr_SI        -3623.750000 -5152.177734 -3362.491699 -4231.451172   \n",
       "ctr_VEE       -3776.950684 -5093.940918 -3489.823242 -4465.648438   \n",
       "ctr_pre_AMB   -3575.081055 -5142.391602 -3328.959717 -4333.919922   \n",
       "ctr_OEE_1     -3746.059326 -5191.430664 -3425.160156 -4422.537109   \n",
       "ctr_AMB_dist  -4184.358887 -5787.214844 -3891.475586 -4949.999023   \n",
       "ctr_AMB_prox  -4025.751953 -5646.947754 -3745.275391 -4811.572754   \n",
       "ctr_ISR_SI    -3731.051270 -5124.352051 -3417.758301 -4320.359375   \n",
       "\n",
       "                      8            9        ...              3334  \\\n",
       "ctr_DEEx      -3743.825195 -3645.994629     ...      -3718.858398   \n",
       "ctr_upper_IEE -3733.900879 -3626.227539     ...      -3714.743164   \n",
       "ctr_OEE_IEE   -3771.521729 -3695.071777     ...      -3721.506104   \n",
       "ctr_OEE_2     -3836.349854 -3766.377686     ...      -3784.227539   \n",
       "ctr_OSR       -3801.904053 -3734.900635     ...      -3771.967529   \n",
       "ctr_SI        -3749.849609 -3664.914307     ...      -3748.192139   \n",
       "ctr_VEE       -3860.351074 -3780.692383     ...      -3832.022949   \n",
       "ctr_pre_AMB   -3689.512207 -3595.837646     ...      -3756.101562   \n",
       "ctr_OEE_1     -3849.865967 -3806.821045     ...      -3823.564941   \n",
       "ctr_AMB_dist  -4254.697266 -4150.666504     ...      -4347.254883   \n",
       "ctr_AMB_prox  -4079.221680 -3988.664795     ...      -4210.827148   \n",
       "ctr_ISR_SI    -3805.827881 -3747.845215     ...      -3810.211914   \n",
       "\n",
       "                      3335         3336         3337         3338  \\\n",
       "ctr_DEEx      -4774.810547 -3405.502441 -3209.570312 -3864.764893   \n",
       "ctr_upper_IEE -4835.724609 -3373.500000 -3220.977295 -3836.332520   \n",
       "ctr_OEE_IEE   -4784.722656 -3469.545166 -3257.449463 -3906.003662   \n",
       "ctr_OEE_2     -4701.400391 -3548.579102 -3312.280762 -3954.685791   \n",
       "ctr_OSR       -4659.078613 -3516.998535 -3315.391846 -3826.553467   \n",
       "ctr_SI        -4702.145996 -3454.336426 -3279.175781 -3783.816895   \n",
       "ctr_VEE       -4754.228516 -3553.413574 -3355.746582 -3998.401367   \n",
       "ctr_pre_AMB   -4803.732422 -3413.587402 -3270.857422 -3883.302246   \n",
       "ctr_OEE_1     -4748.683594 -3615.694824 -3367.099365 -4038.809570   \n",
       "ctr_AMB_dist  -5462.891602 -3977.195801 -3789.163086 -4452.683594   \n",
       "ctr_AMB_prox  -5303.000488 -3820.992676 -3653.866211 -4309.155273   \n",
       "ctr_ISR_SI    -4591.441406 -3556.550537 -3340.044434 -3847.442627   \n",
       "\n",
       "                      3339         3340         3341         3342         3343  \n",
       "ctr_DEEx      -3724.276855 -4436.709473 -4342.840820 -4401.244629 -4121.519531  \n",
       "ctr_upper_IEE -3723.338379 -4413.879883 -4315.059082 -4414.622070 -4197.270508  \n",
       "ctr_OEE_IEE   -3803.436523 -4337.062500 -4443.498047 -4507.806152 -4243.020020  \n",
       "ctr_OEE_2     -3856.008545 -4407.008789 -4508.538086 -4547.631836 -4244.314453  \n",
       "ctr_OSR       -3841.211670 -4255.573242 -4489.912598 -4537.487305 -4261.460938  \n",
       "ctr_SI        -3782.928955 -4315.301758 -4397.071289 -4464.516113 -4236.380859  \n",
       "ctr_VEE       -3882.810303 -4483.126953 -4555.071289 -4568.892578 -4286.059570  \n",
       "ctr_pre_AMB   -3745.017822 -4435.564453 -4301.487305 -4366.380371 -4199.173828  \n",
       "ctr_OEE_1     -3922.755127 -4444.573242 -4587.330078 -4635.699219 -4315.931152  \n",
       "ctr_AMB_dist  -4317.268066 -5040.771484 -4979.303223 -5000.744629 -4838.551270  \n",
       "ctr_AMB_prox  -4161.617676 -4914.133789 -4790.520996 -4803.109375 -4669.141113  \n",
       "ctr_ISR_SI    -3855.660156 -4313.973633 -4508.205078 -4543.085449 -4237.670410  \n",
       "\n",
       "[12 rows x 3344 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# log likelihoods. Can be use directly to find maximum likelihood.\n",
    "bay"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['ctr_pre_AMB_like', 'ctr_upper_IEE_like', 'ctr_upper_IEE_like',\n",
       "       ..., 'ctr_pre_AMB_like', 'ctr_pre_AMB_like', 'ctr_DEEx_like'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# maximum likelihood for every cell\n",
    "recreated = bay.idxmax().apply(lambda x: x+'_like').values\n",
    "recreated"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compare to results provided on GEO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "on_geo = adata[toclassify].obs['class2_3_like'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 out of 3344 cells have mismatching labels\n"
     ]
    }
   ],
   "source": [
    "mismatches = recreated!=on_geo\n",
    "print(\"%d out of %d cells have mismatching labels\"%(mismatches.sum(),len(mismatches)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
